src/sparselm/model/_stepwise.py

```diff
@@ -0,0 +1,164 @@
+
+from itertools import chain
+
+import numpy as np
+from numpy.typing import ArrayLike
+from sklearn.base import RegressorMixin
+from sklearn.linear_model._base import LinearModel
+from sklearn.utils.metaestimators import _BaseComposition
+
+# BaseComposition makes sure that StepwiseEstimator can be correctly cloned.
+
+
+class StepwiseEstimator(_BaseComposition, RegressorMixin, LinearModel):
+    """A composite estimator used to do stepwise fitting.
+
+    The first estimator in the composite will be used to fit
+    certain features (a piece of the feature matrix) to the
+    target vector, and the residuals are fitted to the rest
+    of features by using the next estimators in the composite.
+    """
+
+    def __init__(self, steps, estimator_feature_indices):
+        """Initialize estimator.
+
+        Args:
+            steps(list[(str, CVXEstimator)]):
+                A list of step names and the CVXEstimators to use
+                for each step. CompositeEstimator cannot be used as
+                a member of CompositeEstimator.
+                An estimator will fit the residuals of the previous
+                estimator fits in the list.
+            estimator_feature_indices(list[list[int]]):
+                Scope of each estimator, which means the indices of
+                features in the scope (features[:, scope]) will be
+                fitted to the residual using the corresponding estimator.
+                Notice:
+                   If estimators in the composite requires hierarchy
+                   or groups, the indices in the groups or hierarchy
+                   must be adjusted such that they correspond to the groups
+                   or hierarchy relations in the part of features sliced
+                   by scope.
+                   For example, consider original groups = [0, 1, 1, 2, 2],
+                   and an estimator has scope = [3, 4], then the estimator
+                   should be initialized with group = [0, 0].
+                   You are fully responsible to initialize the estimators
+                   with correct hierarchy, groups and other parameters before
+                   wrapping them up with the composite!
+        """
+        if len(steps) != len(estimator_feature_indices):
+            raise ValueError("Must specify a list of feature indices" " for each step!")
+        full_scope = sorted(set(chain(*estimator_feature_indices)))
+        if full_scope != sorted(
+            chain(*estimator_feature_indices)
+        ) or full_scope != list(range(len(full_scope))):
+            raise ValueError(
+                f"Estimator feature indices: {estimator_feature_indices}"
+                f" can not overlap and must be continuous!"
+            )
+        for _, estimator in steps:
+            if isinstance(estimator, StepwiseEstimator):
+                raise ValueError(
+                    "Cannot add a StepwiseEstimator into a" " CompositeEstimator!"
+                )
```

Comment:
parameter checks should not be done in `__init__` see guidelines here:

https://scikit-learn.org/stable/developers/develop.html